document/js-api/index.bs

```diff
@@ -1115,6 +1147,111 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
+        1. Let |args| be the result of [=coerce JavaScript arguments|coercing arguments=] (|functype|,|arguments|)
+        1. Let (|store|, |result|) be the result of [=func_invoke=](|store|, |funcaddr|, |args|).
+        1. Assert: if control reaches here, we have done waiting for suspended imports
+        1. If |result|.\[[Type]] is NORMAL,
+            1. Perform [=EvaluateCall=] (|promise|.\[[Resolve]],|result|.\[[Value]],false)
+            1. Return UNDETERMINED
+        1. If |result|.\[[Type]] is THROW,
+            1. Perform [=EvaluateCall=] (|promise|.\[[Reject]],|result|.\[[Value]],false).
+            1. Return UNDETERMINED
+        1. Assert fatal error
+    1. Let |con| be [=CreateBuiltinFunction=](|runner|,3,"",[])
+    1. Perform [$Call$](|con|, undefined, [])
+    1. Returns |promise|
+</div>
+
+Note: The extra |$Call$| in the above algorithm ensures that the creation of the Promise is separated from the fullfilling of that Promise. In effect, this allows suspension of the fullfillment to occur whilst allowing the creation of the Promise itself to continue.
+
+<div algorithm>
+    The <dfn constructor for="Suspending">Suspending(|jsFun|)</dfn> constructor, when invoked, performs the following steps:
+    1. If [$IsCallable$](|jsFun|) is false, throw a {{TypeError}}.
+    1. Let |suspendingProto| be \[[%WebAssembly.Suspending.prototype%]]
+    1. Let |susp| be the result of [$OrdinaryObjectCreate$](|suspendingProto|)
+    1. Assign the \[[wrappedFunction]] internal slot of |susp| to |jsFun|
+    1. Return |susp|
+</div>
+
+<div algorithm>
+To <dfn>create a suspending function</dfn> from a JavaScript function |func|, with type |functype| perform the following steps:
+
+    1. Assert: [$IsCallable$](|func|).
+    1. Let |stored settings| be the <a spec=HTML>incumbent settings object</a>.
+    1. Let |hostfunc| be a [=host function=] which performs the following steps when called with arguments |arguments|:
+        1. Let |realm| be |func|'s [=associated Realm=].
+        1. Let |relevant settings| be |realm|'s [=realm/settings object=].
+        1. Let |async_context| be the [=surrounding agent=]'s [=running execution context=].
+        1. [=Prepare to run script=] with |relevant settings|.
+        1. [=Prepare to run a callback=] with |stored settings|.
+        1. Let [|parameters|] → [|results|] be |functype|.
+        1. Let |jsArguments| be the result of [=coerce WebAssembly arguments=](|parameters|,|arguments|)
+        1. Let |ret| be [=?=] [$Call$](|func|, undefined, |jsArguments|).
+        1. [=Clean up after running a callback=] with |stored settings|.
+        1. [=Clean up after running script=] with |relevant settings|.
+        1. Assert: |ret|.\[[Type]] is <emu-const>throw</emu-const> or <emu-const>normal</emu-const>.
+        1. If |ret|.\[[Type]] is <emu-const>throw</emu-const>, then trigger a WebAssembly trap, and propagate |ret|.\[[Value]] to the enclosing JavaScript.
```

Comment:
We should remember to update this when rebasing on top of exception handling. Would it make sense to do that rebase sooner rather than later?

Reply:
The actual state of the js-promising repo is a bit 'rough'.
In any case, we should rebase on top of exception handling. Propose doing that after this pr lands on the js-promising repo

Reply:
The spec/wasm-3.0 branch now contains EH. If you want to minimise later merge conflicts, I'd suggest directly rebasing on that.